### PR TITLE
Adds Syncthing as a data mover in VolSync

### PR DESCRIPTION
**Pull request by osilkin98**
**Describe what this PR does**

**Is there anything that requires special attention?**

**Related issues:**

### CODE REVIEW

File: mover-syncthing/config.xml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    </folder>
+    <gui enabled="true" tls="true" debugging="false">
+        <address>127.0.0.1:8384</address>
+        <theme>default</theme>
+    </gui>
+    <ldap></ldap>
+    <options>
+        <listenAddress>default</listenAddress>
+        <globalAnnounceServer>default</globalAnnounceServer>
+        <globalAnnounceEnabled>true</globalAnnounceEnabled>
```
Comment (JohnStrunk) [inline mover-syncthing/config.xml:50]: I bet we could disable this since we have to provide IPs manually anyway. Folks would probably appreciate not having the connection to the discovery servers.

File: mover-syncthing/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,85 @@
+# Build the Syncthing image from Golang
+FROM golang:1.17 AS syncthing-builder
```
Comment (JohnStrunk) [inline mover-syncthing/Dockerfile:2]:
```suggestion
FROM registry.access.redhat.com/ubi8/go-toolset as builder
```
... to be consistent w/ the other movers

File: mover-syncthing/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  https://github.com/syncthing/syncthing.git
+WORKDIR /workspace/syncthing
+
+# Make sure we have the correct Syncthing release
+RUN /bin/bash -c "[[ $(git rev-list -n 1 HEAD) == ${SYNCTHING_GIT_HASH} ]]"
+# Create the binary for syncthing 
+RUN go run build.go
+
+
+ENTRYPOINT [ "/workspace/syncthing/bin/syncthing" ]
```
Comment (JohnStrunk) [inline mover-syncthing/Dockerfile:25]: Don't need an entrypoint in the builder
```suggestion
```

File: mover-syncthing/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+RUN git clone \
+  --depth 1 \
+  --branch=${SYNCTHING_VERSION} \
+  https://github.com/syncthing/syncthing.git
+WORKDIR /workspace/syncthing
+
+# Make sure we have the correct Syncthing release
+RUN /bin/bash -c "[[ $(git rev-list -n 1 HEAD) == ${SYNCTHING_GIT_HASH} ]]"
+# Create the binary for syncthing 
+RUN go run build.go
```
Comment (JohnStrunk) [inline mover-syncthing/Dockerfile:22]: I think we want to disable autoupgrade. I think that amounts to adding `-no-upgrade` to this command

File: mover-syncthing/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+# RUN useradd -ms /bin/bash stuser
+# USER root
+
+RUN microdnf update -y && \
+    microdnf clean all && \
+    rm -rf /var/cache/yum
+
+EXPOSE 21027/udp 2000/tcp 22000 8384/tcp
+
+COPY --from=syncthing-builder \
```
Comment (JohnStrunk) [inline mover-syncthing/Dockerfile:39]: To match the L2 change
```suggestion
COPY --from=builder \
```

File: mover-syncthing/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+      org.label-schema.vcs-url="https://github.com/backube/volsync" \
+      org.label-schema.vendor="Backube" \
+      org.label-schema.version="${version}"
+###### syncthing-specific args ############
+
+# Path where the config volume will be mounted
+ENV SYNCTHING_CONFIG_DIR="/config"
+
+# Bind the GUI address to 0.0.0.0 so we can get external requests
+ENV STGUIADDRESS=0.0.0.0:8384
```
Comment (JohnStrunk) [inline mover-syncthing/Dockerfile:66]: Is this overriding the setting of gui.address on L43 of the config file? Should we just set it there?

File: mover-syncthing/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+# Variables that will be used to configure the data volume on startup
+# folder path 
+ENV SYNCTHING_DATA_DIR="/data"
+# folder id - must be the same on all synced nodes
+ENV SYNCTHING_DATA_ID="my-folder-id" 
+# folder label
+ENV SYNCTHING_DATA_LABEL="synced volume"
```
Comment (JohnStrunk) [inline mover-syncthing/Dockerfile:74]: Is there a reason to customize these or could we just hardcode them in the config file?

File: mover-syncthing/.dockerignore [hunk 8]
```diff
@@ -0,0 +1 @@
+./deployments/*
```
Comment (JohnStrunk) [inline mover-syncthing/.dockerignore:1]: Is this file needed? Perhaps it's a testing leftover?

**Review by JohnStrunk — COMMENTED** [review]
- Please ensure the files have newlines at the end
- This also needs updates to the .github/workflows/operator.yml file to build and push the new container.

File: mover-syncthing/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+  --depth 1 \
+  --branch=${SYNCTHING_VERSION} \
+  https://github.com/syncthing/syncthing.git
+WORKDIR /workspace/syncthing
+
+# Make sure we have the correct Syncthing release
+RUN /bin/bash -c "[[ $(git rev-list -n 1 HEAD) == ${SYNCTHING_GIT_HASH} ]]"
+# Create the binary for syncthing 
+RUN go run build.go -no-upgrade
+
```
Comment (JohnStrunk) [inline mover-syncthing/Dockerfile:23]:
```suggestion

# Verify that FIPS crypto libs are accessible
RUN nm /workspace/syncthing/bin/syncthing | grep -q goboringcrypto
```

**Review by JohnStrunk — COMMENTED** [review]
Sorry, I missed the `nm` bit in my initial review.

Thanks for adding the build to the workflow. We also need a job in the workflow to push the built container (like the other movers).

File: mover-syncthing/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+  --depth 1 \
+  --branch=${SYNCTHING_VERSION} \
+  https://github.com/syncthing/syncthing.git
+WORKDIR /workspace/syncthing
+
+# Make sure we have the correct Syncthing release
+RUN /bin/bash -c "[[ $(git rev-list -n 1 HEAD) == ${SYNCTHING_GIT_HASH} ]]"
+
+# Ensure that syncthing builds with boringcryptolib 
+ENV GOLANG_FIPS=1
```
Comment (JohnStrunk) [inline mover-syncthing/Dockerfile:23]: Is this needed?
My understand from reading https://developers.redhat.com/blog/2019/06/24/go-and-fips-140-2-on-red-hat-enterprise-linux#fips_140_2_validated_cryptography is that this would be used at runtime to force the use of fips libraries as opposed to autodetecting based on the system's FIPS mode.

Comment (RobotSail) [reply mover-syncthing/Dockerfile]: Nope, I just removed it.

Comment (RobotSail) [reply mover-syncthing/Dockerfile]: Nope, I just removed it
